rep-2011.rst

```diff
@@ -321,7 +321,214 @@ The passed ``rosidl_message_type_support_t`` in the init call can be introspecte
+
+      ...
+    }
+
+Plugin Deserialization
+""""""""""""""""""""""
+
+One way a serialization library plugin for FastCDR could be implemented could be as follows, relying on the serialization library to deserialize the message buffer, but with the provided type description to supply type reflection information:
+
+.. code::
+
+    #include <FastBuffer.h>
+    #include <FastCdr.h>
+
+    // If the parsed description is iterable
+    // rcl_serialization_plugin_t->(*deserialize)
+    void deserialize(void *buffer,
+                     plugin_internal_description_t *description,
+                     rcl_deserialized_message_t *msg)
+    {
+      FastCdr * cdr = new FastCdr(new FastBuffer((char *)buffer, description->buffer_length));
+
+      msg->message_field_names = new char*[description->num_fields];
+      msg->message_types = new char*[description->num_fields];
+
+      for (int i = 0; i < description->num_fields; i++)
+      {
+        const char * type = description->field_types[i];
+
+        // Primitives
+        if (strcmp(type, "bool") == 0)
+        {
+          bool tmp;
+        }
+        else if (strcmp(type, "char") == 0)
+        {
+          char tmp;
+        }
+        ...
+
+        // TODO:: What to do about nested types? A recursive unpacking call with a different description?
+
+        cdr >> tmp;
+        msg->message_fields[description->field_names[i]] = reinterpret_cast<void *>(tmp);
+
+        msg->message_field_names[i] = description->field_names[i];
+        msg->message_types[i] = description->field_types[i];
+      }
+      msg->message_field_count = description->num_fields;
+    }
+
+Example Introspection API
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once the serialization library plugins are able to deserialize the raw message buffer, downstream programs can then introspect the constructed deserialized message object, which should be laid out as such:
+
+.. code::
+
+    struct rcl_deserialized_field_t {
+        void * value;
+        char * type;
+    }
+
+    struct rcl_deserialized_message_t {
+        int message_field_count;
+        const char** message_field_names;
+        const char** message_types;
+
+        // Some dynamically allocated key->value associative map type storing void * field values
+        rcl_associative_array message_fields;
+
+        // Function pointers
+        rcl_deserialized_field_t * (*get_field_by_index)(int index);
+    };
+
+Now, for a given message description `Foo.msg`:
+
+.. code::
+
+    // Foo.msg
+    bool bool_field
+    char char_field
+    float32 float_field
+
+The corresponding `rcl_deserialized_message_t` can be queried accordingly:
+
+.. code::
+
+    rcl_deserialized_message_t * foo_msg;
+    foo_msg->message_field_names[0];  // "bool_field"
+    foo_msg->message_types[0];  // "bool"
+
+    // Get the field
+    if (strcmp("bool", foo_msg->get_field_by_index(0)->type) == 0)
+    {
+        *((bool*)foo_msg->get_field_by_index(0)->value);
+    }
```

Comment:
All this needs to be updated with the nested types work.

Reply:
Will be updated in the next PR